Circuit relay v1

lchenut opened the pull request:
A simple way to make it work at the moment:
```nim
proc main() {.async, gcsafe.} =
  let
    rng = newRng()
    src = newStandardSwitch(rng=rng)
    rel = newStandardSwitch(rng=rng)
    address = MultiAddress.init("/ip4/127.0.0.1/tcp/0").tryGet()
    addressRelay = MultiAddress.init("/ip4/127.0.0.1/tcp/0/p2p-circuit").tryGet()
    dst = SwitchBuilder
      .new()
      .withRng(rng)
      .withAddresses(@[ address, addressRelay ])
      .withTcpTransport()
      .withMplex()
      .withNoise()
      .build()
    pingProtocol = Ping.new(rng=rng)
    relaySrc = Relay.new(src, false)
    relayRel = Relay.new(rel, true)
    relayDst = Relay.new(dst, false)

  src.mount(relaySrc)
  rel.mount(relayRel)
  dst.mount(relayDst)
  dst.mount(pingProtocol)
  src.transports &= RelayTransport.new(relaySrc)
  ((Dialer)src.dialer).transports = src.transports
  dst.transports &= RelayTransport.new(relayDst)
  ((Dialer)dst.dialer).transports = dst.transports

  let
    srcFut = await src.start()
    relFut = await rel.start()
    dstFut = await dst.start()

  let ma = MultiAddress.init($rel.peerInfo.addrs[0] & "/p2p/" & $rel.peerInfo.peerId & "/p2p-circuit/p2p/" & $dst.peerInfo.peerId).tryGet()
  await src.connect(rel.peerInfo.peerId, rel.peerInfo.addrs)
  await rel.connect(dst.peerInfo.peerId, dst.peerInfo.addrs)
  let conn = await src.dial(dst.peerInfo.peerId, @[ ma ], PingCodec)

  echo "ping: ", await pingProtocol.ping(conn)

  await allFutures(src.stop(), rel.stop(), dst.stop())
  await allFutures(srcFut & relFut & dstFut)
waitFor(main())
```
Few notes:
- The destination having two multiaddresses is clearly undesirable but it's a temporary workaround because without circuit-relay address, the transport won't start. 
- Speaking of temporary workaround, Dial.transports becoming public is one.
- It may be interesting to create an access control list as it is implemented in [the go-libp2p version](https://github.com/libp2p/go-libp2p/blob/master/p2p/protocol/circuitv1/relay/relay.go#L133-L137) even though there's no mention of it in the specs. It's relatively easy to implement.

dryajov commented on libp2p/protocols/relay.nim:
```diff
@@ -0,0 +1,471 @@
+    r.hopCount.inc(src.peerId, -1)
+    r.hopCount.inc(dst.peerId, -1)
+
+  if r.streamCount > r.maxCircuit:
```
This should probably be at the top, before we parse the message and all that.

dryajov commented on libp2p/protocols/relay.nim:
```diff
@@ -0,0 +1,471 @@
+# dstPeerId = QmTarget
+# dstAddrs = /ipv4/127.0.0.1/tcp/5003
+#
+# TODO: Must find a way to solve the handles problem, currently the dst switch is created
```
Not exactly sure what the problem is, it's should be totally valid to start with the `/p2p-circuit` address, that might be another way of announcing circuit capabilities in the future.

Menduist replied:
Actually, I just re-read this, and I think the issue is that you shouldn't supply the addresses manually at start, since you don't know which relay will relay to you

So I think we shouldn't have any p2p-circuit address in our multiaddress list (until we handle the reservations from circuit relay v2), and should work as-is?

dryajov replied:
Ah, I see - yes, we might allow an empty/no addresses at start, that shouldn't be a big issue.

dryajov commented on libp2p/protocols/relay.nim:
```diff
@@ -0,0 +1,471 @@
+  result = await self.dial(address)
+
+method handles*(self: RelayTransport, ma: MultiAddress): bool {.gcsafe} =
+  result=false
```
`bools` are false by default

dryajov commented on libp2p/protocols/relay.nim:
```diff
@@ -0,0 +1,471 @@
+  trace "Handles return", ma, result
+
+proc new*(T: typedesc[RelayTransport], relay: Relay): T =
+  T(relay: relay, upgrader: relay.switch.transports[0].upgrader)
```
There isn't any need to do this, you can safelly override `upgradeIncoming` and `upgradeOutgoing` to do nothing and it would work as expected.

Menduist replied:
Edit: we actually need to upgrade, though adding support in `builders.nim` is a better solution

dryajov replied:
Ah, you're right, we need to upgrade the connection and correct, we can just pass it as a parameter. 

I wonder if we can pass the `transports` array as well without having to expose it in the switch - we should be able to too.

dryajov reviewed (changes requested):
Hey @lchenut! Thank you so much for taking the time to work on this! Great job and appologies for taking this long to respond from my side!

So far everything looks pretty good sans a few minor things I've already commented on. In any case I'll give it another pass soon.

One crussial thing that is missing are interop tests. We have examples of how we do this in the `testinterop.nim` file, we run them against a go libp2p daemon instance. It's very usefull in catching any inconsistencies and validate how well supported the protocol. Let me know if you need any help navigating it.

Cheers!

dryajov commented on tests/testrelay.nim:
```diff
@@ -0,0 +1,242 @@
+    pingProtocol {.threadVar.}: Ping
+
+  proc createMsg(
+    msgType: Option[RelayType] = none[RelayType](),
```
The idiomatic way to writ this would be `msgType = RelayType.none`

Menduist commented on libp2p/protocols/relay.nim:
```diff
@@ -0,0 +1,471 @@
+    msgTypeOrd: uint32
+    src: RelayPeer
+    dst: RelayPeer
+    status: RelayStatus
```
Unused?

Menduist commented on libp2p/protocols/relay.nim:
```diff
@@ -0,0 +1,471 @@
+    dst: RelayPeer
+    status: RelayStatus
+    statusOrd: uint32
+    pb = initProtoBuffer(buf)
```
Can be moved to the let

Menduist commented on libp2p/protocols/relay.nim:
```diff
@@ -0,0 +1,471 @@
+proc sendStatus*(conn: Connection, code: RelayStatus) {.async, gcsafe.} =
+  trace "send status", status = $code & "(" & $ord(code) & ")"
+
+  var
```
Can be a let

Menduist commented on libp2p/protocols/relay.nim:
```diff
@@ -0,0 +1,471 @@
+  #       and not r.acl.AllowHop(src.peerId, dst.peerId)
+  #         sendStatus(conn, RelayStatus.HopCantSpeakRelay)
+
+  r.streamCount += 1
```
```suggestion
  r.streamCount.inc()
```

Menduist commented on libp2p/protocols/relay.nim:
```diff
@@ -0,0 +1,471 @@
+  r.hopCount.inc(dst.peerId)
+  defer:
+    r.streamCount -= 1
+    r.hopCount.inc(src.peerId, -1)
```
There should be a dec, no?

Menduist commented on libp2p/protocols/relay.nim:
```diff
@@ -0,0 +1,471 @@
+  await conn.join()
+
+proc handleCanHop(r: Relay, conn: Connection, msg: RelayMessage) {.async, gcsafe.} =
+  if r.canHop:
```
```nim
await sendStatus(conn,
  if r.canHop:
    RelayStatus.Success
  else:
    RelayStatus.HopCantSpeakRelay
)
```

Less code duplication

Menduist commented on libp2p/protocols/relay.nim:
```diff
@@ -0,0 +1,471 @@
+
+const
+  CircuitRelay = mapEq("p2p-circuit")
+  P2PPat = mapEq("p2p")
```
We can move that to multiaddress.nim

Menduist commented on libp2p/protocols/relay.nim:
```diff
@@ -0,0 +1,471 @@
+    dstPeerId: PeerId
+  if not relayPeerId.init(($(sma[^3].get())).split('/')[2]):
+    # error
+    discard
```
This will not do anything? Better to do something with the error

Menduist reviewed (commented):
Some nitpicks

lchenut commented on libp2p/protocols/relay.nim:
```diff
@@ -0,0 +1,471 @@
+
+  trace "relaying connection", src, dst
+
+  var
```
The buffers have to be initialized outside the loop, the reads that have a reference to these buffers are outside the scope of the loop. The best I can do for the moment is to create a separate bridge procedure, to use the buffers outside the main function.

Eventually, a good solution could be to use (if it exists) or create (if not) a bridge function from chronos, but I'm scared of this library right now.

lchenut commented:
@dryajov @Menduist, I corrected all the errors / nitpicks and I added the interop tests. The only thing I left open is the dirty buffers in the bridge part of the relay as I wanted your advice on this point.
The new way to make it works could be this:
```nim
proc main() {.async, gcsafe.} =
  let
    CustomProtoCodec = "/customProto"
    rng = newRng()
    src = SwitchBuilder.new()
      .withRng(rng)
      .withAddresses(@[ MultiAddress.init("/ip4/127.0.0.1/tcp/0").tryGet() ])
      .withTcpTransport()
      .withMplex()
      .withNoise()
      .withRelayTransport(canHop=false)
      .build()
    rel = SwitchBuilder.new()
      .withRng(rng)
      .withAddresses(@[ MultiAddress.init("/ip4/127.0.0.1/tcp/0").tryGet() ])
      .withTcpTransport()
      .withMplex()
      .withNoise()
      .withRelayTransport(canHop=true)
      .build()
    dst = SwitchBuilder.new()
      .withRng(rng)
      .withAddresses(@[ MultiAddress.init("/ip4/127.0.0.1/tcp/0").tryGet() ])
      .withTcpTransport()
      .withMplex()
      .withNoise()
      .withRelayTransport(canHop=false)
      .build()

  # Create a custom protocol
  var proto = new LPProtocol
  proto.handler = proc (conn: Connection, customProto: string) {.async.} =
    echo "Destination Read From Source: ", string.fromBytes(await conn.readLP(1024))
    await conn.writeLp("I'm out")
    await conn.close()
  proto.codec = CustomProtoCodec

  dst.mount(proto)

  await src.start()
  await rel.start()
  await dst.start()

  let ma = MultiAddress.init($rel.peerInfo.addrs[0] & "/p2p/" & $rel.peerInfo.peerId & "/p2p-circuit/p2p/" & $dst.peerInfo.peerId).tryGet()
  # /ip4/.../tcp/.../p2p/QmRelay/p2p-circuit/p2p/QmDestination
  await src.connect(rel.peerInfo.peerId, rel.peerInfo.addrs)
  await rel.connect(dst.peerInfo.peerId, dst.peerInfo.addrs)
  let conn = await src.dial(dst.peerInfo.peerId, @[ ma ], CustomProtoCodec)

  await conn.writeLp("Hello")
  echo "Source Read From Destination: ", string.fromBytes(await conn.readLp(1024))

  await allFutures(src.stop(), rel.stop(), dst.stop())
waitFor(main())
```

Few more points:
- [This test](https://github.com/lchenut/nim-libp2p/blob/f6d6a5c630d645615d5ba6f7948b4dfaa9c88d03/tests/testrelay.nim#L290-L340) is really redundant with the previous one, I was originally only testing the withRelayTransport of the builder and decide to keep it (more tests is (nearly) always better)
- Speaking of the builder, [this part](https://github.com/lchenut/nim-libp2p/blob/f6d6a5c630d645615d5ba6f7948b4dfaa9c88d03/libp2p/builders.nim#L208-L212) is really dirty (still cleaner than the old version though) and hiding the Relay behind the switch prevents us to change the three configurable variables (number maximum of circuit, number maximum of circuit per peer and message size)
- There's always this ACL thing, not mentioned in the spec, that [the go version does](https://github.com/libp2p/go-libp2p/blob/0e72b4832bb3949d15291aa0ba7883040691a24a/p2p/protocol/circuitv1/relay/relay.go#L180-L184)
- [This is my bad, I fix it now](https://github.com/lchenut/nim-libp2p/blob/f6d6a5c630d645615d5ba6f7948b4dfaa9c88d03/libp2p/protocols/relay.nim#L489-L493)

dryajov commented on libp2p/protocols/relay.nim:
```diff
@@ -0,0 +1,491 @@
+  # TODO: Eventually create a similar procedure somewhere in the chronos library
+  proc bridge(conn: Connection, connDst: Connection) {.async.} =
+    var
+      bufSrcToDst: array[4096, byte]
```
I would use the `{.noInit.}` pragma here and then `zeroMem` every time I finished using the buffer in the loop.

Menduist replied:
Not sure I understand the issue here, this code can't leak any data unless `read` returns a wrong value, which would be catastrophic anyway
And even if a bit of the buffer was somehow reused, that would be in the same relayed connection, so it would only resend some old data to the same peer

And even if all of that happened, since the connection from A to B through R is secured anyway, you wouldn't be able to extract any useful data? R could post the entire exchange on Facebook and nothing bad would happen

The only thing we have to be careful about is to not leak our internal memory, so we must be careful to zero mem the buffers before any usage (which is done implicitly right now)

lchenut replied:
After a little thought and more deeper inspection on how pipes work kernel-side (with a buffer, more or less the same thing I did) I must agree with @Menduist, there's not really a better solution.

A `const bufferSize = 4096` or `bufSrcToDst.high + 1` are good ways to avoid the repeat of the 4096 that could be problematic if someone updates this section of the code.

And I understand the use of the `zeroMem` at the end of the loop, even though, like Menduist pointed out, the connection is secured so it's a bit superfluous, but we never know what can happen. The `{.noInit.}` pragma on another hand is less effective to secure this part, if the code fails during the first loop, it could be dangerous with the buffer uninitialized.

lchenut commented:
@dryajov @Menduist I fixed the few minor errors I mentioned at the last meeting. If you could review the code that would be great.

Menduist commented on libp2p/protocols/relay.nim:
```diff
@@ -0,0 +1,500 @@
+    r3 = pb.getField(3, pbDst)
+    r4 = pb.getField(4, statusOrd)
+
+  res = r1.isOk() and r2.isOk() and r3.isOk() and r4.isOk()
```
Why not

```nim
if not res: return none(RelayMessage)
```

And the follow ifs can be just `if r2.get()`

And no need for the last one

Menduist replied:
Actually, I would do something like:
```nim
let res = r1.isOk() and r2.isOk() and r3.isOk() and r4.isOk()
if not res: return none(RelayMessage)

if r2.get() and
   pbSrc.getField(1, sec.peerId) == false or
   pbSrc.getRepeatedField(2, sec.addrs) == false:
     return none(RelayMessage)

[..]
some(
```
Or something similar

lchenut replied:
To be fair, it was one of the first function I wrote and I was (maybe?) influenced by the code of "identify" at that time.
I could change it, but I'll change the identify decodeMsg function aswell.

Menduist commented on libp2p/protocols/relay.nim:
```diff
@@ -0,0 +1,500 @@
+  let
+    msg = RelayMessage(
+      msgType: some(RelayType.Status),
+      srcPeer: none(RelayPeer),
+      dstPeer: none(RelayPeer),
```
none is the default value, no need to specify them I think

Menduist commented on libp2p/protocols/relay.nim:
```diff
@@ -0,0 +1,500 @@
+    await sendStatus(conn, RelayStatus.HopCantSpeakRelay)
+    return
+
+  let connDst = try:
+    await r.switch.dial(dst.peerId, @[RelayCodec])
+  except CancelledError as exc:
+    raise exc
+  except CatchableError as exc:
+    trace "error opening relay stream", dst, exc=exc.msg
+    await sendStatus(conn, RelayStatus.HopCantDialDst)
+    return
+  defer:
+    await connDst.close()
+
+  let msgToSend = RelayMessage(
+    msgType: some(RelayType.Stop),
+    srcPeer: some(src),
+    dstPeer: some(dst),
+    status: none(RelayStatus))
+  try:
+    await connDst.writeLp(encodeMsg(msgToSend).buffer)
+  except CancelledError as exc:
+    raise exc
+  except CatchableError as exc:
+    trace "error writing stop handshake", exc=exc.msg
+    await sendStatus(conn, RelayStatus.HopCantOpenDstStream)
+    return
+
+  let msgRcvFromDstOpt = try:
+    decodeMsg(await connDst.readLp(r.msgSize))
+  except CancelledError as exc:
+    raise exc
+  except CatchableError as exc:
+    trace "error reading stop response", exc=exc.msg
+    await sendStatus(conn, RelayStatus.HopCantOpenDstStream)
+    return
```
Maybe this could be a single try / except, instead of 3

lchenut replied:
The last two try were merged in a previous version, I separated them just to make the trace more accurate.

Menduist commented on libp2p/protocols/relay.nim:
```diff
@@ -0,0 +1,500 @@
+    return
+
+  try:
+    await sendStatus(conn, RelayStatus.Success)
```
Why is this sendStatus in a try / except and not the other ones?

Menduist commented on libp2p/protocols/relay.nim:
```diff
@@ -0,0 +1,500 @@
+        trace "relay handle stream", msg = msgOpt.get()
+      let msg = msgOpt.get()
+
+      if msg.msgType.get() == RelayType.Hop:
```
Can't use a case?

```nim
case msg.msgType.get():
of Hop: ###
```

Menduist commented on libp2p/protocols/relay.nim:
```diff
@@ -0,0 +1,500 @@
+  await procCall Transport(self).start(ma)
+  trace "Starting Relay transport"
+
+method stop*(self: RelayTransport) {.async, gcsafe.} =
+  self.running = false
+  while not self.relay.queue.empty():
+    await self.relay.queue.popFirstNoWait().close()
```
This looks a bit risky: a peer might "slip in" during this procedure

Could you block the relay somehow before closing the connections?

Actually, it might make sense to move the `queue` to the transport, and add a callback to the Relay instead.
If the callback is nil, the relay refuses the connection, and the callback would be set to nil at the start of stop

And the transport start would hook in the callback to the transport queues

Menduist commented on libp2p/protocols/relay.nim:
```diff
@@ -0,0 +1,500 @@
+method accept*(self: RelayTransport): Future[Connection] {.async, gcsafe.} =
+  result = await self.relay.queue.popFirst()
+
+# TODO: dial / handles active relay:
+# /ipv4/127.0.0.1/tcp/5002/p2p/QmRelay/p2p-circuit/ipv4/127.0.0.1/tcp/5003/p2p/QmTarget
+# dstPeerId = QmTarget
+# dstAddrs = /ipv4/127.0.0.1/tcp/5003
```
As discussed somewhere, the active relay has been deprecated, so we can safely remove this

Menduist commented on libp2p/protocols/relay.nim:
```diff
@@ -0,0 +1,500 @@
+proc dial*(self: RelayTransport, ma: MultiAddress): Future[Connection] {.async, gcsafe.} =
+  let
+    sma = toSeq(ma.items())
+    relayAddrs = sma[0..sma.len-4].mapIt(it.tryGet()).foldl(a & b)
```
Careful when using this kind of operator, they can raise a KeyError which would crash the entire program
Could you add some tests with invalid multiaddress, to be sure this is handled properly

Menduist commented on libp2p/builders.nim:
```diff
@@ -197,6 +205,11 @@ proc build*(b: SwitchBuilder): Switch
     ms = ms,
     nameResolver = b.nameResolver)
 
+  if b.isCircuitRelay:
+    let relay = Relay.new(switch, b.circuitRelayCanHop)
+    switch.mount(relay)
+    switch.addTransport(RelayTransport.new(relay, muxedUpgrade))
```
Actually, now that this is cleaner, we could even remove it from the builder, and let the user build it himself (like gossipsub)

Menduist reviewed (commented):
Getting close!

dryajov reviewed (approved):
Great job @lchenut! LGTM!
